### PR TITLE
Alyapunov/gh 6206 repeat read violation v2

### LINKED ISSUE #6206: mvcc: repeatable read violation after delete
**Version**
```console
$ tarantool --version

Tarantool 2.9.0-119-g29e2931c6
Target: Linux-x86_64-Debug
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/cc /usr/bin/c++
C_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -Werror
CXX_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -Werror
```

**Reproducer (requires `txn_proxy.lua` from `test/box/lua`)**
```lua
yaml = require 'yaml'

box.cfg{memtx_use_mvcc_engine=true}

txn_proxy = require('txn_proxy')

spc = box.schema.space.create('test')
idx = spc:create_index('test', {type='tree'})

tx1 = txn_proxy.new()

tx1:begin()
print(yaml.encode(tx1('spc:select{}')))
spc:replace{1, 1}
print(yaml.encode(tx1('spc:select{}')))
tx1:commit()

spc:delete{1}

tx2 = txn_proxy.new()

tx2:begin()
print(yaml.encode(tx2('spc:select{}')))
spc:replace{1, 1}
print(yaml.encode(tx2('spc:select{}')))
tx2:commit()

spc:drop()
```

**Actual behavior**
```console
---
- []
...

---
- []
...

---
- []
...

---
- [[1, 1]]
...
```

**Expected behavior**
```console
---
- []
...

---
- []
...

---
- []
...

---
- []
...
```

### PR DESCRIPTION
**Pull request by alyapunov**
In order to preserve repeated reads transactional manager tracks
read of each transactions. Generally reads can be of two types -
those that have read a tuple or that have found nothing. The first
are stored in tuple story, the second - in special gap and hole
structures.
    
The problem was that reads that found a dirty tuple that was
invisible to this transaction (the story says that it is deleted)
was not stored neither in story nor in gap/holes.
    
This patch fixes that.

Closes #6206

### CODE REVIEW

File: src/box/memtx_tx.c [hunk 1]
```diff
@@ -1944,6 +1922,42 @@ memtx_tx_track_read(struct txn *txn, struct space *space, struct tuple *tuple)
...
+		return 0;
+	if (space == NULL)
+		return 0;
+	if (space->def->opts.is_ephemeral)
+		return 0;
+
+	if (tuple->is_dirty) {
+		struct memtx_story *story = memtx_tx_story_get(tuple);
+		return memtx_tx_track_read_story(txn, space, story);
+	} else {
```
Comment (EvgenyMekhanik) [inline src/box/memtx_tx.c:1974]: Here we can remove 'else', because we return in 'if' part. Up to you.

**Comment by EvgenyMekhanik** [comment]
There is a memory leak in `memtx_tx_track_gap_slow` (this does not related to this patch, but I saw it during the review process): story = memtx_tx_story_new(space, successor); if `story == NULL` we not truncate region from `item->key`.

**Review by EvgenyMekhanik — APPROVED** [review]

File: src/box/memtx_tx.h [hunk 2]
```diff
@@ -75,6 +75,8 @@ struct tx_read_tracker {
 	struct rlist in_reader_list;
 	/** Link in reader->read_set. */
 	struct rlist in_read_set;
+	/** Bit field of indexes in which the data was tread by reader. */
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply src/box/memtx_tx.h:78]: What we gonna do if the number of indexes is greater than 64?:)

**Review by Korablev77 — COMMENTED** [review]
Changelog is missing...

**Review by Korablev77 — APPROVED** [review]
Otherwise seems good

**Comment by Korablev77** [comment]
Already pushed to master.

**Pull request closed without merging**